File: rudder-unity-client/sdk/rudder-unity-sdk/Event/Property/TrackPropertyBuilder.cs
﻿using System;
namespace com.rudderlabs.unity.library.Event.Property
{
    public class TrackPropertyBuilder : RudderPropertyBuilder
    {
        private string category;
        public TrackPropertyBuilder SetCategory(string category)
        {
            this.category = category;
            return this;
        }

        private string label;
        public TrackPropertyBuilder SetLabel(string label)
        {
            this.label = label;
            return this;
        }

        private string value;
        public TrackPropertyBuilder SetValue(string value)
        {
            this.value = value;
            return this;
        }

        public override RudderProperty Build()
        {
            if (category == null)
            {
                throw new Exception("Key \"category\" is required for track event");
            }

            RudderProperty rudderProperty = new RudderProperty();
            rudderProperty.AddProperty("category", this.category);
            if (label != null)
            {
                rudderProperty.AddProperty("label", this.label);
            }
            if (value != null)
            {
                rudderProperty.AddProperty("value", this.value);
            }

            return rudderProperty;
        }
    }
}

File: rudder-client-android-kotlin-old/src/main/java/com/rudderlabs/android/library/models/porperties/TrackPropertyBuilder.kt
package com.rudderlabs.android.library.models.porperties

import com.rudderlabs.android.library.exceptions.MalformedEventException

class TrackPropertyBuilder : RudderPropertyBuilder() {
    private var category: String? = null
    fun setCategory(category: String): TrackPropertyBuilder {
        this.category = category
        return this
    }

    private var label: String = ""
    fun setLabel(label: String): TrackPropertyBuilder {
        this.label = label
        return this
    }

    private var value: String = ""
    fun setValue(value: String): TrackPropertyBuilder {
        this.value = value
        return this
    }

    override fun build(): RudderProperty {
        if (category == null)
            throw MalformedEventException("Key \"category\" is required for track event")

        return RudderProperty().also {
            it.addProperty("category", category)
            it.addProperty("label", label)
            it.addProperty("value", value)
        }
    }
}
